src/hotspot/share/cds/archiveBuilder.cpp

```diff
@@ -791,67 +815,116 @@ void ArchiveBuilder::make_klasses_shareable() {
         type = "plat";
-        num_platform_klasses ++;
+        ADD_COUNT(num_platform_klasses);
       } else if (ik->is_shared_app_class()) {
         type = "app";
-        num_app_klasses ++;
+        ADD_COUNT(num_app_klasses);
       } else {
         assert(ik->is_shared_unregistered_class(), "must be");
         type = "unreg";
-        num_unregistered_klasses ++;
+        ADD_COUNT(num_unregistered_klasses);
+      }
+
+      if (AOTClassLinker::is_vm_class(src_ik)) {
+        ADD_COUNT(num_vm_klasses);
       }
 
       if (!ik->is_linked()) {
-        num_unlinked_klasses ++;
-        unlinked = " ** unlinked";
+        ADD_COUNT(num_unlinked_klasses);
+        unlinked = " unlinked";
+        if (ik->is_shared_boot_class()) {
+          boot_unlinked ++;
+        } else if (ik->is_shared_platform_class()) {
+          platform_unlinked ++;
+        } else if (ik->is_shared_app_class()) {
+          app_unlinked ++;
+        } else {
+          unreg_unlinked ++;
+        }
       }
 
       if (ik->is_hidden()) {
-        num_hidden_klasses ++;
-        hidden = " ** hidden";
+        ADD_COUNT(num_hidden_klasses);
+        hidden = " hidden";
       }
 
       if (ik->is_generated_shared_class()) {
-        generated = " ** generated";
+        generated = " generated";
+      }
+      if (aotlinked) {
+        aotlinked_msg = " aot-linked";
       }
+
       MetaspaceShared::rewrite_nofast_bytecodes_and_calculate_fingerprints(Thread::current(), ik);
       ik->remove_unshareable_info();
     }
 
     if (log_is_enabled(Debug, cds, class)) {
       ResourceMark rm;
-      log_debug(cds, class)("klasses[%5d] = " PTR_FORMAT " %-5s %s%s%s%s", i,
+      log_debug(cds, class)("klasses[%5d] = " PTR_FORMAT " %-5s %s%s%s%s%s", i,
                             p2i(to_requested(k)), type, k->external_name(),
-                            hidden, unlinked, generated);
+                            hidden, unlinked, generated, aotlinked_msg);
     }
   }
 
+#define STATS_FORMAT    "= %5d, aot-linked = %5d"
+#define STATS_PARAMS(x) num_ ## x, num_ ## x ## _a
+
   log_info(cds)("Number of classes %d", num_instance_klasses + num_obj_array_klasses + num_type_array_klasses);
-  log_info(cds)("    instance classes   = %5d", num_instance_klasses);
-  log_info(cds)("      boot             = %5d", num_boot_klasses);
-  log_info(cds)("      app              = %5d", num_app_klasses);
-  log_info(cds)("      platform         = %5d", num_platform_klasses);
-  log_info(cds)("      unregistered     = %5d", num_unregistered_klasses);
-  log_info(cds)("      (hidden)         = %5d", num_hidden_klasses);
-  log_info(cds)("      (unlinked)       = %5d", num_unlinked_klasses);
+  log_info(cds)("    instance classes   " STATS_FORMAT, STATS_PARAMS(instance_klasses));
+  log_info(cds)("      boot             " STATS_FORMAT, STATS_PARAMS(boot_klasses));
+  log_info(cds)("       vm              " STATS_FORMAT, STATS_PARAMS(vm_klasses));
```

Comment:
```suggestion
  log_info(cds)("      vm              " STATS_FORMAT, STATS_PARAMS(vm_klasses));
```

Reply:
The indentation is intentional: vm is a subset of boot classes, which is a subset of instance classes.

Reply:
Okay but I presume the indent level should be the same: boot was indented by 2, then vm only be 1.